### PR TITLE
Coverity fixes for walreceiver.c and xlog_mm.c

### PR DESCRIPTION
**Pull request by asubramanya**
CID 129947: Out-of-bounds write (OVERRUN)
overrun-local: Overrunning array buf of 80 bytes at byte offset 80 using index
nbytes (which evaluates to 80).
The above defect was found in `WalRcvUsr2Handler()` which is in gpdb only and
not in upstream.

CID 129271, 129270, 129269, 129268 : Function address comparison (BAD_COMPARE)
func_conv: This implicit conversion to a function pointer is suspicious:
The above defect was found in `WalReceiverMain()`. The comparison with NULL was
being performed against a function name and not a function pointer or variable
which is clearly incorrect. Since we do static linking of the functions, the
check was not necessary in the first place. Fixed the defect by removing the
check.

 CID 129288, 129277:
 Buffer not null terminated (BUFFER_SIZE_WARNING)
 buffer_size_warning: Calling strncpy with a maximum size argument of 1024 bytes
 on destination array m->path2 of size 1024 bytes might leave the destination
 string unterminated.
 This issue was detected in the `xlog_mm_redo()` function. Fixed by using
 `strlcpy()` to ensure that the buffer is null terminated.
    
 CID 129856: Argument cannot be negative (NEGATIVE_RETURNS)
 negative_returns: fd is passed to a parameter that cannot be negative.
 This issue was also found in the `xlog_mm_redo()` function.
 Fixed the issue by making sure that `close()` is not called on negative fd.

### CODE REVIEW

File: src/backend/replication/walreceiver.c [hunk 1]
```diff
@@ -272,11 +272,6 @@ WalReceiverMain(void)
 	/* We allow SIGQUIT (quickdie) at all times */
 	sigdelset(&BlockSig, SIGQUIT);
 
-	/* Load the libpq-specific functions */
-	if (walrcv_connect == NULL || walrcv_receive == NULL ||
```
Comment (asimrp) [inline src/backend/replication/walreceiver.c]: It makes sense to remove the check as they are statically defined but these are function pointers and not strings as the commit message suggests.

Comment at the top of `gp_libpqwalreceiver.c` is not valid, as far as I can tell.  The comment says that the file is built as a dynamic library, loaded at run time.  The build steps in Makefile don't indicate that.  Am I missing something?

File: src/backend/access/transam/xlog_mm.c [hunk 2]
```diff
@@ -182,10 +182,10 @@ mmxlog_redo(XLogRecPtr beginLoc, XLogRecPtr lsn, XLogRecord *record)
 			m.filespaceoid = xlrec->filespace;
 
 			m.dbid1 = xlrec->u.dbid.master;
-    		strncpy(m.path1, xlrec->master_path, MAXPGPATH);
+			strlcpy(m.path1, xlrec->master_path, MAXPGPATH);
```
Comment (asimrp) [inline src/backend/access/transam/xlog_mm.c]: I'm assuming a filespace path that's longer than MAXPGPATH is already rejected / truncated much before it makes it into xlog record.  So it is not necessary to check if `xlrec->master_path` was truncated by `strlcpy()`.

Comment (ashwinstar) [inline src/backend/access/transam/xlog_mm.c]: May not be necessary but no harm as well in using safer version `strlcpy()` instead.

File: src/backend/replication/walreceiver.c (hunk 1 above)
Comment (ashwinstar) [inline src/backend/replication/walreceiver.c]: Comment is wrong, its statically linked. We will fix the the comment.

**Comment by asubramanya** [comment]
Thanks for the review.

**Pull request closed without merging**